src/submodule.c

```diff
@@ -149,6 +149,46 @@ static int find_by_path(const git_config_entry *entry, void *payload)
 	return 0;
 }
 
+/*
+ * Checks to see if the submodule shares its name with a file or directory that
+ * already exists on the index. If so, the submodule cannot be added.
+ */
+static int can_add_submodule(git_repository *repo, const char *path)
```

Comment:
I think this is a bit misnamed. You do not check if we can add the submodule at a given path, as that would involve other checks as well. Instead, you only check whether the given path is already occupied by something else. How about `is_path_occupied` instead?

One more thing is that functions which should answer a certain question are expected to return a bool, not an int. So I'd propose a signature like `static int is_path_occupied(bool *occupied, git_repository *repo, const char *path)`. Like this, a real unexpected error can be returned by returing a negative value. If no error happens we return `0` and `occupied` will then be set to `true` if the path is occupied and `false` otherwise.